File: c/anharmonic/phonon3/collision_matrix.c
/* Copyright (C) 2015 Atsushi Togo */
/* All rights reserved. */

/* This file is part of phonopy. */

/* Redistribution and use in source and binary forms, with or without */
/* modification, are permitted provided that the following conditions */
/* are met: */

/* * Redistributions of source code must retain the above copyright */
/*   notice, this list of conditions and the following disclaimer. */

/* * Redistributions in binary form must reproduce the above copyright */
/*   notice, this list of conditions and the following disclaimer in */
/*   the documentation and/or other materials provided with the */
/*   distribution. */

/* * Neither the name of the phonopy project nor the names of its */
/*   contributors may be used to endorse or promote products derived */
/*   from this software without specific prior written permission. */

/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS */
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT */
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS */
/* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE */
/* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, */
/* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, */
/* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; */
/* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER */
/* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT */
/* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN */
/* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE */
/* POSSIBILITY OF SUCH DAMAGE. */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <phonoc_array.h>
#include <phonoc_utils.h>
#include <phonon3_h/collision_matrix.h>

static void get_collision_matrix(double *collision_matrix,
                                 const double *fc3_normal_squared,
                                 const size_t num_band0,
                                 const size_t num_band,
                                 const double *frequencies,
                                 const size_t (*triplets)[3],
                                 const size_t *triplets_map,
                                 const size_t num_gp,
                                 const size_t *map_q,
                                 const size_t *rot_grid_points,
                                 const size_t num_ir_gp,
                                 const size_t num_rot,
                                 const double *rotations_cartesian,
                                 const double *g,
                                 const double temperature,
                                 const double unit_conversion_factor,
                                 const double cutoff_frequency);
static void
get_reducible_collision_matrix(double *collision_matrix,
                               const double *fc3_normal_squared,
                               const size_t num_band0,
                               const size_t num_band,
                               const double *frequencies,
                               const size_t (*triplets)[3],
                               const size_t *triplets_map,
                               const size_t num_gp,
                               const size_t *map_q,
                               const double *g,
                               const double temperature,
                               const double unit_conversion_factor,
                               const double cutoff_frequency);
static void get_inv_sinh(double *inv_sinh,
                         const size_t gp,
                         const double temperature,
                         const double *frequencies,
                         const size_t triplet[3],
                         const size_t *triplets_map,
                         const size_t *map_q,
                         const size_t num_band,
                         const double cutoff_frequency);
static size_t *create_gp2tp_map(const size_t *triplets_map,
                                const size_t num_gp);

void col_get_collision_matrix(double *collision_matrix,
                              const Darray *fc3_normal_squared,
                              const double *frequencies,
                              const size_t (*triplets)[3],
                              const size_t *triplets_map,
                              const size_t *map_q,
                              const size_t *rot_grid_points,
                              const double *rotations_cartesian,
                              const double *g,
                              const size_t num_ir_gp,
                              const size_t num_gp,
                              const size_t num_rot,
                              const double temperature,
                              const double unit_conversion_factor,
                              const double cutoff_frequency)
{
  size_t num_triplets, num_band0, num_band;

  num_triplets = fc3_normal_squared->dims[0];
  num_band0 = fc3_normal_squared->dims[1];
  num_band = fc3_normal_squared->dims[2];

  get_collision_matrix(
    collision_matrix,
    fc3_normal_squared->data,
    num_band0,
    num_band,
    frequencies,
    triplets,
    triplets_map,
    num_gp,
    map_q,
    rot_grid_points,
    num_ir_gp,
    num_rot,
    rotations_cartesian,
    g + 2 * num_triplets * num_band0 * num_band * num_band,
    temperature,
    unit_conversion_factor,
    cutoff_frequency);
}

void col_get_reducible_collision_matrix(double *collision_matrix,
                                        const Darray *fc3_normal_squared,
                                        const double *frequencies,
                                        const size_t (*triplets)[3],
                                        const size_t *triplets_map,
                                        const size_t *map_q,
                                        const double *g,
                                        const size_t num_gp,
                                        const double temperature,
                                        const double unit_conversion_factor,
                                        const double cutoff_frequency)
{
  size_t num_triplets, num_band, num_band0;

  num_triplets = fc3_normal_squared->dims[0];
  num_band0 = fc3_normal_squared->dims[1];
  num_band = fc3_normal_squared->dims[2];

  get_reducible_collision_matrix(
    collision_matrix,
    fc3_normal_squared->data,
    num_band0,
    num_band,
    frequencies,
    triplets,
    triplets_map,
    num_gp,
    map_q,
    g + 2 * num_triplets * num_band0 * num_band * num_band,
    temperature,
    unit_conversion_factor,
    cutoff_frequency);
}

static void get_collision_matrix(double *collision_matrix,
                                 const double *fc3_normal_squared,
                                 const size_t num_band0,
                                 const size_t num_band,
                                 const double *frequencies,
                                 const size_t (*triplets)[3],
                                 const size_t *triplets_map,
                                 const size_t num_gp,
                                 const size_t *map_q,
                                 const size_t *rot_grid_points,
                                 const size_t num_ir_gp,
                                 const size_t num_rot,
                                 const double *rotations_cartesian,
                                 const double *g,
                                 const double temperature,
                                 const double unit_conversion_factor,
                                 const double cutoff_frequency)
{
  size_t i, j, k, l, m, n, ti, r_gp;
  size_t *gp2tp_map;
  double collision;
  double *inv_sinh;

  gp2tp_map = create_gp2tp_map(triplets_map, num_gp);

#pragma omp parallel for private(j, k, l, m, n, ti, r_gp, collision, inv_sinh)
  for (i = 0; i < num_ir_gp; i++) {
    inv_sinh = (double*)malloc(sizeof(double) * num_band);
    for (j = 0; j < num_rot; j++) {
      r_gp = rot_grid_points[i * num_rot + j];
      ti = gp2tp_map[triplets_map[r_gp]];
      get_inv_sinh(inv_sinh,
                   r_gp,
                   temperature,
                   frequencies,
                   triplets[ti],
                   triplets_map,
                   map_q,
                   num_band,
                   cutoff_frequency);

      for (k = 0; k < num_band0; k++) {
        for (l = 0; l < num_band; l++) {
          collision = 0;
          for (m = 0; m < num_band; m++) {
            collision +=
              fc3_normal_squared[ti * num_band0 * num_band * num_band +
                                 k * num_band * num_band +
                                 l * num_band + m] *
              g[ti * num_band0 * num_band * num_band +
                k * num_band * num_band +
                l * num_band + m] *
              inv_sinh[m] * unit_conversion_factor;
          }
          for (m = 0; m < 3; m++) {
            for (n = 0; n < 3; n++) {
              collision_matrix[k * 3 * num_ir_gp * num_band * 3 +
                               m * num_ir_gp * num_band * 3 +
                               i * num_band * 3 + l * 3 + n] +=
                collision * rotations_cartesian[j * 9 + m * 3 + n];
            }
          }
        }
      }
    }
    free(inv_sinh);
    inv_sinh = NULL;
  }

  free(gp2tp_map);
  gp2tp_map = NULL;
}

static void
get_reducible_collision_matrix(double *collision_matrix,
                               const double *fc3_normal_squared,
                               const size_t num_band0,
                               const size_t num_band,
                               const double *frequencies,
                               const size_t (*triplets)[3],
                               const size_t *triplets_map,
                               const size_t num_gp,
                               const size_t *map_q,
                               const double *g,
                               const double temperature,
                               const double unit_conversion_factor,
                               const double cutoff_frequency)
{
  size_t i, j, k, l, ti;
  size_t *gp2tp_map;
  double collision;
  double *inv_sinh;

  gp2tp_map = create_gp2tp_map(triplets_map, num_gp);

#pragma omp parallel for private(j, k, l, ti, collision, inv_sinh)
  for (i = 0; i < num_gp; i++) {
    inv_sinh = (double*)malloc(sizeof(double) * num_band);
    ti = gp2tp_map[triplets_map[i]];
    get_inv_sinh(inv_sinh,
                 i,
                 temperature,
                 frequencies,
                 triplets[ti],
                 triplets_map,
                 map_q,
                 num_band,
                 cutoff_frequency);

    for (j = 0; j < num_band0; j++) {
      for (k = 0; k < num_band; k++) {
        collision = 0;
        for (l = 0; l < num_band; l++) {
          collision +=
            fc3_normal_squared[ti * num_band0 * num_band * num_band +
                               j * num_band * num_band +
                               k * num_band + l] *
            g[ti * num_band0 * num_band * num_band +
              j * num_band * num_band +
              k * num_band + l] *
            inv_sinh[l] * unit_conversion_factor;
        }
        collision_matrix[j * num_gp * num_band + i * num_band + k] += collision;
      }
    }

    free(inv_sinh);
    inv_sinh = NULL;
  }

  free(gp2tp_map);
  gp2tp_map = NULL;
}

static void get_inv_sinh(double *inv_sinh,
                         const size_t gp,
                         const double temperature,
                         const double *frequencies,
                         const size_t triplet[3],
                         const size_t *triplets_map,
                         const size_t *map_q,
                         const size_t num_band,
                         const double cutoff_frequency)
{
  size_t i, gp2;
  double f;

  if (triplets_map[gp] == map_q[gp]) {
    gp2 = triplet[2];
  } else {
    gp2 = triplet[1];
  }
  for (i = 0; i < num_band; i++) {
    f = frequencies[gp2 * num_band + i];
    if (f > cutoff_frequency) {
      inv_sinh[i] = inv_sinh_occupation(f, temperature);
    } else {
      inv_sinh[i] = 0;
    }
  }
}

static size_t *create_gp2tp_map(const size_t *triplets_map,
                                const size_t num_gp)
{
  size_t i, max_i, count;
  size_t *gp2tp_map;

  max_i = 0;
  for (i = 0; i < num_gp; i++) {
    if (max_i < triplets_map[i]) {
      max_i = triplets_map[i];
    }
  }

  gp2tp_map = (size_t*)malloc(sizeof(size_t) * (max_i + 1));
  for (i = 0; i < max_i + 1; i++) {
    gp2tp_map[i] = 0;
  }

  count = 0;
  for (i = 0; i < num_gp; i++) {
    if (triplets_map[i] == i) {
      gp2tp_map[i] = count;
      count++;
    }
 }

  return gp2tp_map;
}

File: phono3py/phonon3/collision_matrix.py
"""Calculate collision matrix of direct solution of LBTE."""

# Copyright (C) 2020 Atsushi Togo
# All rights reserved.
#
# This file is part of phono3py.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in
#   the documentation and/or other materials provided with the
#   distribution.
#
# * Neither the name of the phonopy project nor the names of its
#   contributors may be used to endorse or promote products derived
#   from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from typing import Optional

import numpy as np
from phonopy.units import Kb, THzToEv

from phono3py.phonon3.imag_self_energy import ImagSelfEnergy
from phono3py.phonon3.interaction import Interaction


class CollisionMatrix(ImagSelfEnergy):
    """Collision matrix of direct solution of LBTE for one grid point.

    Main diagonal part (imag-self-energy) and
    the other part are separately stored.

    """

    def __init__(
        self,
        interaction: Interaction,
        rotations_cartesian=None,
        num_ir_grid_points=Optional[int],
        rot_grid_points=None,
        is_reducible_collision_matrix=False,
        log_level=0,
        lang="C",
    ):
        """Init method."""
        self._pp: Interaction
        self._is_collision_matrix: bool
        self._sigma = None
        self._frequency_points = None
        self._temperature = None
        self._grid_point = None
        self._lang = None
        self._imag_self_energy = None
        self._collision_matrix = None
        self._pp_strength = None
        self._frequencies = None
        self._triplets_at_q = None
        self._triplets_map_at_q = None
        self._weights_at_q = None
        self._band_indices = None
        self._unit_conversion = None
        self._cutoff_frequency = None
        self._g = None
        self._unit_conversion = None
        self._log_level = log_level

        super().__init__(interaction, lang=lang)

        self._is_reducible_collision_matrix = is_reducible_collision_matrix
        self._is_collision_matrix = True

        if not self._is_reducible_collision_matrix:
            self._num_ir_grid_points = num_ir_grid_points
            self._rot_grid_points = np.array(
                self._pp.bz_grid.bzg2grg[rot_grid_points], dtype="int_", order="C"
            )
            self._rotations_cartesian = rotations_cartesian

    def run(self):
        """Calculate collision matrix at a grid point."""
        if self._pp_strength is None:
            self.run_interaction()

        num_band0 = self._pp_strength.shape[1]
        num_band = self._pp_strength.shape[2]
        self._imag_self_energy = np.zeros(num_band0, dtype="double")

        if self._is_reducible_collision_matrix:
            num_mesh_points = np.prod(self._pp.mesh_numbers)
            self._collision_matrix = np.zeros(
                (num_band0, num_mesh_points, num_band), dtype="double"
            )
        else:
            self._collision_matrix = np.zeros(
                (num_band0, 3, self._num_ir_grid_points, num_band, 3), dtype="double"
            )
        self._run_with_band_indices()
        self._run_collision_matrix()

    def get_collision_matrix(self):
        """Return collision matrix at a grid point."""
        return self._collision_matrix

    def set_grid_point(self, grid_point=None):
        """Set a grid point and prepare for collision matrix calculation."""
        if grid_point is None:
            self._grid_point = None
        else:
            self._pp.set_grid_point(grid_point, store_triplets_map=True)
            self._pp_strength = None
            (
                self._triplets_at_q,
                self._weights_at_q,
                self._triplets_map_at_q,
                self._ir_map_at_q,
            ) = self._pp.get_triplets_at_q()
            self._grid_point = grid_point
            self._frequencies, self._eigenvectors, _ = self._pp.get_phonons()

    def _run_collision_matrix(self):
        if self._temperature > 0:
            if self._lang == "C":
                if self._is_reducible_collision_matrix:
                    self._run_c_reducible_collision_matrix()
                else:
                    self._run_c_collision_matrix()
            else:
                if self._is_reducible_collision_matrix:
                    self._run_py_reducible_collision_matrix()
                else:
                    self._run_py_collision_matrix()

    def _run_c_collision_matrix(self):
        import phono3py._phono3py as phono3c

        phono3c.collision_matrix(
            self._collision_matrix,
            self._pp_strength,
            self._frequencies,
            self._g,
            self._triplets_at_q,
            self._triplets_map_at_q,
            self._ir_map_at_q,
            self._rot_grid_points,  # in GRGrid
            self._rotations_cartesian,
            self._temperature,
            self._unit_conversion,
            self._cutoff_frequency,
        )

    def _run_c_reducible_collision_matrix(self):
        import phono3py._phono3py as phono3c

        phono3c.reducible_collision_matrix(
            self._collision_matrix,
            self._pp_strength,
            self._frequencies,
            self._g,
            self._triplets_at_q,
            self._triplets_map_at_q,
            self._ir_map_at_q,
            self._temperature,
            self._unit_conversion,
            self._cutoff_frequency,
        )

    def _run_py_collision_matrix(self):
        r"""Sum over rotations, and q-points and bands for third phonons.

        \Omega' = \sum_R' R' \Omega_{kp,R'k'p'}

        pp_strength.shape = (num_triplets, num_band0, num_band, num_band)

        """
        num_band0 = self._pp_strength.shape[1]
        num_band = self._pp_strength.shape[2]
        gp2tp, tp2s, swapped = self._get_gp2tp_map()
        for i in range(self._num_ir_grid_points):
            r_gps = self._rot_grid_points[i]
            for r, r_gp in zip(self._rotations_cartesian, r_gps):
                inv_sinh = self._get_inv_sinh(tp2s[r_gp])
                ti = gp2tp[r_gp]
                for j, k in np.ndindex((num_band0, num_band)):
                    if swapped[r_gp]:
                        collision = (
                            self._pp_strength[ti, j, :, k]
                            * inv_sinh
                            * self._g[2, ti, j, :, k]
                        ).sum()
                    else:
                        collision = (
                            self._pp_strength[ti, j, k]
                            * inv_sinh
                            * self._g[2, ti, j, k]
                        ).sum()
                    collision *= self._unit_conversion
                    self._collision_matrix[j, :, i, k, :] += collision * r

    def _run_py_reducible_collision_matrix(self):
        r"""Sum over q-points and bands of third phonons.

        This corresponds to the second term of right hand side of
        \Omega_{q0p0, q1p1} in Chaput's paper.

        pp_strength.shape = (num_triplets, num_band0, num_band, num_band)

        """
        num_mesh_points = np.prod(self._pp.mesh_numbers)
        num_band0 = self._pp_strength.shape[1]
        num_band = self._pp_strength.shape[2]
        gp2tp, tp2s, swapped = self._get_gp2tp_map()
        for gp1 in range(num_mesh_points):
            inv_sinh = self._get_inv_sinh(tp2s[gp1])
            ti = gp2tp[gp1]
            for j, k in np.ndindex((num_band0, num_band)):
                if swapped[gp1]:
                    collision = (
                        self._pp_strength[ti, j, :, k]
                        * inv_sinh
                        * self._g[2, ti, j, :, k]
                    ).sum()
                else:
                    collision = (
                        self._pp_strength[ti, j, k] * inv_sinh * self._g[2, ti, j, k]
                    ).sum()
                collision *= self._unit_conversion
                self._collision_matrix[j, gp1, k] += collision

    def _get_gp2tp_map(self):
        """Return mapping table from grid point index to triplet index.

        triplets_map_at_q contains index mapping of q1 in (q0, q1, q2) to
        independet q1 under q0+q1+q2=G with a fixed q0.

        Note
        ----
        map_q[gp1] <= gp1.:
            Symmetry relation of grid poi nts with a stabilizer q0.
        map_triplets[gp1] <= gp1 :
            map_q[gp1] == gp1 : map_q[gp2] if map_q[gp2] < gp1 otherwise gp1.
            map_q[gp1] != gp1 : map_triplets[map_q[gp1]]



        As a rule
        1. map_triplets[gp1] == gp1 : [gp0, gp1, gp2]
        2. map_triplets[gp1] != gp1 : [gp0, map_q[gp2], gp1'],
                                      map_triplets[gp1] == map_q[gp2]

        """
        map_triplets = self._triplets_map_at_q
        map_q = self._ir_map_at_q
        gp2tp = -np.ones(len(map_triplets), dtype="int_")
        tp2s = -np.ones(len(map_triplets), dtype="int_")
        swapped = np.zeros(len(map_triplets), dtype="bytes")
        num_tps = 0

        bzg2grg = self._pp.bz_grid.bzg2grg

        for gp1, tp_gp1 in enumerate(map_triplets):
            if map_q[gp1] == gp1:
                if gp1 == tp_gp1:
                    gp2tp[gp1] = num_tps
                    tp2s[gp1] = self._triplets_at_q[num_tps][2]
                    assert bzg2grg[self._triplets_at_q[num_tps][1]] == gp1
                    num_tps += 1
                else:  # q1 <--> q2 swap if swappable.
                    gp2tp[gp1] = gp2tp[tp_gp1]
                    tp2s[gp1] = self._triplets_at_q[gp2tp[gp1]][1]
                    swapped[gp1] = 1
                    assert map_q[bzg2grg[self._triplets_at_q[gp2tp[gp1]][2]]] == gp1
            else:  # q1 is not in ir-q1s.
                gp2tp[gp1] = gp2tp[map_q[gp1]]
                tp2s[gp1] = tp2s[map_q[gp1]]
                swapped[gp1] = swapped[map_q[gp1]]

            # Alternative implementation of tp2s
            # grg2bzg = self._pp.bz_grid.grg2bzg
            # addresses = self._pp.bz_grid.addresses
            # q0 = addresses[self._triplets_at_q[0][0]]
            # q1 = addresses[grg2bzg[gp1]]
            # q2 = -q0 - q1
            # gp2 = get_grid_point_from_address(q2, self._pp.bz_grid.D_diag)
            # tp2s[gp1] = self._pp.bz_grid.grg2bzg[gp2]

        return gp2tp, tp2s, swapped

    def _get_inv_sinh(self, gp):
        """Return sinh term for bands at a q-point."""
        freqs = self._frequencies[gp]
        sinh = np.where(
            freqs > self._cutoff_frequency,
            np.sinh(freqs * THzToEv / (2 * Kb * self._temperature)),
            -1.0,
        )
        inv_sinh = np.where(sinh > 0, 1.0 / sinh, 0)

        return inv_sinh
